File: 125. Valid Palindrome/python/125_valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        stripped_str = ''.join(ch.lower() for ch in s if ch.isalnum())
        length = len(stripped_str)
        if length==0:
            return True
        
        if length%2!=0:
            front = back = int(length/2)
        else:
            front = int(length/2)-1
            back = int(front + 1)

        while (front >=0 and back < length and stripped_str[front]==stripped_str[back]):
            front -=1
            back +=1
            
        if front == -1 and back == length:
            return True
        
        return False
        
sol = Solution()

test1 = "A man, a plan, a canal: Panama"
test2 = "race a car"
test3 = " "


print(sol.isPalindrome(test1))
print(sol.isPalindrome(test2))
print(sol.isPalindrome(test3))

File: Algorithms/125. Valid Palindrome/go/125_valid_palindrome.go
package main

import (
	"fmt"
	"strings"
)

func isPalindrome(s string) bool {
	i, j := 0, len(s)-1
	for i < j {
		if !isValid(s[i]) {
			i++
			continue
		}
		if !isValid(s[j]) {
			j--
			continue
		}
		if !strings.EqualFold(string(s[i]), string(s[j])) {
			return false
		}
		i++
		j--
	}

	return true
}

func isValid(a byte) bool {
	if (a >= 'a' && a <= 'z') || (a >= 'A' && a <= 'Z') || (a >= '0' && a <= '9') {
		return true
	}
	return false
}

func main() {
	test1 := "A man, a plan, a canal: Panama"
	test2 := "race a car"
	test3 := " "

	fmt.Println(isPalindrome(test1))
	fmt.Println(isPalindrome(test2))
	fmt.Println(isPalindrome(test3))
}
